src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+        {
+            var nameToParameterDict = new Dictionary<TIdentifierNameSyntax, IParameterSymbol>();
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = document.SemanticModel;
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol is IParameterSymbol parameterSymbol)
+                {
+                    nameToParameterDict.Add(variable, parameterSymbol);
+                }
+            }
+
+            return nameToParameterDict;
+        }
+
+        /// <summary>
+        /// Determines if the expression is contained within something that is "parameterized"
+        /// </summary>
+        private static (bool isParameterized, bool hasOptionalParameter) ExpressionWithinParameterizedMethod(
+            SemanticDocument semanticDocument, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var methodSymbol = (IMethodSymbol)semanticDocument.SemanticModel.GetRequiredEnclosingSymbol(expression.SpanStart, cancellationToken);
+
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = semanticDocument.SemanticModel;
+            var hasOptionalParameter = false;
+
+            foreach (var parameter in methodSymbol.Parameters)
+            {
+                if (parameter.HasExplicitDefaultValue)
+                {
+                    hasOptionalParameter = true;
+                }
+            }
+
+            foreach (var variable in variablesInExpression)
+            {
+                var parameterSymbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (parameterSymbol is not IParameterSymbol parameter)
+                {
+                    return (false, hasOptionalParameter);
+                }
+                if (parameter.IsParams)
+                {
+                    return (false, hasOptionalParameter);
+                }
+            }
+
+            return (methodSymbol != null && methodSymbol.GetParameters().Any(), hasOptionalParameter);
+        }
+
+        private ImmutableArray<CodeAction> AddActions(SemanticDocument semanticDocument,
+            TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var actionsBuilder = new ArrayBuilder<CodeAction>();
+            var (isParameterized, hasOptionalParameter) = ExpressionWithinParameterizedMethod(semanticDocument, expression, cancellationToken);
+            if (isParameterized)
+            {
+                actionsBuilder.Add(new MyCodeAction(CreateDisplayText(allOccurrences: false, trampoline: false, overload: false),
+                    c => IntroduceParameterAsync(semanticDocument, expression, false, false, false, cancellationToken)));
```

Comment:
need named parameters for all of this :)

Reply:
you repeat this pattern like 8 times.  extract method or local function to do the work of creating the action.

Reply:
added local function